plugin/src/main/java/io/deephaven/plugin/type/ObjectType.java

```diff
@@ -46,7 +47,85 @@ public interface ObjectType extends Plugin {
+        }
+    }
+
+    /**
+     * A stream of messages, either sent from the server to the client, or client to the server. ObjectType plugin
+     * implementations can provide an implementation of this interface for each incoming stream to invoke as messages
+     * arrive, and will likewise be given an instance of this interface to be able to send messages to the client.
+     */
+    interface MessageStream extends AutoCloseable {
+        /**
+         * Transmits data to the remote end of the stream. This can consist of a binary payload and references to
+         * objects on the server.
+         *
+         * @param payload the binary data sent to the remote implementation
+         * @param references server-side object references sent to the remote implementation
+         */
+        void onMessage(ByteBuffer payload, Object[] references);
+
+        /**
+         * Closes the stream on both ends. No further messages can be sent or received.
+         */
+        void close();
+    }
+
+    /**
+     * Signals creation of a client stream to the provided object. The returned MessageStream implementation will be
+     * called with each received message from the server, and can call the provided connection instance to send messages
+     * as needed to the client.
+     * 
+     * @param object the object to create a connection for
+     * @param connection a stream to send objects to the client
+     * @return a stream to receive objects from the client
+     */
+    // impl note: provide default impl? deprecate writeTo?
+    default MessageStream clientConnection(Object object, MessageStream connection) {
+        if (supportsBidiMessaging(object) == Kind.BIDIRECTIONAL) {
+            // incorrect implementation
+            throw new IllegalStateException(
+                    "ObjectType implementation returned true for supportsBidiMessaging, but has no clientConnection implementation");
+        } else {
+            // internal error, shouldn't have called this
+            throw new IllegalStateException("Do not call clientConnection if supportsBidiMessaging returns false");
+        }
+    }
+
+    /**
+     * Consider renaming to API? MessageAPI?
+     */
+    enum Kind {
+        /**
+         * An object can be fetched, but will not continue to stream results, or receive messages from the server.
+         */
+        FETCHABLE,
+        /**
+         * An object can be connected to by a client, and continue to stream messages server to client or client to
+         * server.
+         */
+        BIDIRECTIONAL,
+    }
+
+    /**
+     * Returns true if the {@code object} supports bidirectional communication.
+     *
+     * @param object the object
+     * @return true if the {@code object} supports bidirectional communication
+     */
+    // TODO review: should the objecttype itself just return fetchable vs bidi? or do we actually want to decide on an
+    // object-by-object basis?
```

Comment:
I think I agree that this should be plugin-level, not per-object. That said, doing it this way preserves optionality for implementers.  Might be confusing, though, for clients.

Reply:
Presently a client doesn't get any say in the matter - and each plugin can only advertise that it handles a single object "type" string, so the type can't be used as a signal for accessing a stream or a single payload. 

However, the initial response payload from the server could include a specific "i'm going to close now" or not, or the server can simply close as a signal "no streams".